scio-neo4j/src/main/scala/com/spotify/scio/neo4j/Neo4jIoOptions.scala

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2022 Spotify AB.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.scio.neo4j
+
+import org.neo4j.driver.{Record, SessionConfig, TransactionConfig}
+
+/**
+ * Options for a Neo4J connection.
+ *
+ * @param url
+ *   connection url, e.g. "neo4j://neo4j.com:7687"
+ * @param username
+ *   login username
+ * @param password
+ *   login password
+ */
+final case class Neo4jConnectionOptions(url: String, username: String, password: String)
+
+sealed trait Neo4jIoOptions
```

Comment:
Shouldn't we define common options here ?

```suggestion
sealed trait Neo4jIoOptions {
  def connectionOptions: Neo4jConnectionOptions
  def cypher: String
}  
```
It will remove the need to pattern match to create the id in `neo4jIoId`

Reply:
Modified in fixup d4cfa0a18abd3c767a741f7a7367f1baaf3ecc48

Reply:
Since this is a sealed trait it does not matter, but in traits, it is preferable to use the least powerful abstraction.
Here `def` instead of `val` so impl can have more freedom when overriding. I think it is also better for binary compatibility too

Reply:
Good catch => e65ed92df37139fbab06f6401ebddc2d7d7732fa